docs/protocol.md

```diff
@@ -1,26 +1,83 @@
-# TChannel v2
+# Specification
 
-## Design goals
+## Introduction
+
+### Problem Statement
+
+TChannel was developed at Uber during a period of explosive growth in its
+[service oriented architecture][soa]. Services living in a large, distributed
+system like this usually encounter similar classes of problems, such as:
+
+ - Service discovery -- how do I discover *and consume* the services around me?
+
+ - Fault tolerance -- how can I best insulate myself from failures outside of
+   my control?
+
+ - [Dapper][dapper]-style tracing -- how can I identify and monitor bottlenecks
+   across the entire system?
+
+[soa]: https://en.wikipedia.org/wiki/Service-oriented_architecture "SOA"
+[dapper]: http://research.google.com/pubs/pub36356.html "Dapper"
+
+### Design Goals
+
+TChannel aims to solve these problems by providing a protocol for clients and
+servers, with an *intelligent routing mesh* (referred to as *Hyperbahn*)
+connecting the two.
+
+Consider those [SOA][soa] problems again:
+
+ - Service discovery: All producers and consumers register themselves with the
+   routing mesh. Consumers access producers only by their name; no need to know
+   about hosts or ports.
+
+ - Fault tolerance: The routing mesh tracks things like failure rates and
+   [SLA][sla] violations. It can intelligently detect unhealthy hosts and
+   remove them from the pool of available hosts.
+
+ - Request tracing: Built into the protocol as a first class citizen.
+
+[sla]: https://en.wikipedia.org/wiki/Service-level_agreement "SLA"
+
+By consolidating logic *between* producers and consumers, this also allows us
+to push core features to our entire SOA without requiring applications to
+update libraries.
+
+Additionally, when developing this protocol and routing mesh, we had the
+following goals in mind:
+
+- The protocol must be easy to implement in multiple languages, especially
+  Javascript, Python, and Go. We do not want to restrict ourselves to one
+  language.
+
+- Async behavior is fundamental. We follow a traditional request / response
+  model with out of order responses; slow requests must not block subsequent,
+  faster requests at the head of the line.
 
-- Easy to implement in multiple languages, especially JS and Python
-- High performance forwarding path. Intermediaries can make a forwarding
-  decision quickly.
-- Request / response model with out of order responses. Slow requests will not
-  block subsequent faster requests at head of line.
 - Large requests/responses may/must be broken into fragments to be sent
-  progressively.
-- Optional checksums
-- Can be used to transport multiple protocols between endpoints, e.g. HTTP+JSON
-  and Thrift.
+  progressively (i.e., "streaming" requests are supported).
+
+- The protocol can be used with multiple transports, e.g. JSON and Thrift,
```

Comment:
s/The protocol/TChannel/

Also, JSON and Thrift are serialization schemes/protocols, not transports.